app/modbus_client.py

```diff
@@ -1,25 +1,26 @@
 from pymodbus.client import ModbusTcpClient
 from app.configuration import Configuration
 
+
 class ModbusClient:
     def __init__(self, configuration: Configuration, port: int, host: str) -> None:
-        self.configuration = configuration 
+        self.configuration = configuration
         self.client = ModbusTcpClient(host, port=port)
 
     def write_coils(self, name: str, value: list[bool]):
         coil_configuration = self.configuration.get_coil(name)
         self.client.connect()
-        self.client.write_coils(coil_configuration.address, value)
+        self.client.write_coils(coil_configuration.address[0], value)
```

Comment:
We are currently manually selecting the first address.
We will need to learn how we handle this when we have multiple addresses, maybe there is a native way to handle this, or maybe we need to manually write to multiple addresses.

Telegraf supports multiple address in its configuration and I think is sensible following the same pattern.

For now though, we are just writing to the first address in the document.

Reply:
pymodbus write_coils here: https://pymodbus.readthedocs.io/en/latest/source/library/client.html#pymodbus.client.mixin.ModbusClientMixin.write_coils

...and the specific implementation here: https://pymodbus.readthedocs.io/en/latest/source/library/pymodbus.html#pymodbus.bit_write_message.WriteMultipleCoilsRequest

This write_coil**s** function is useful only if there is a need to write a sequence of bools to a single address which I don't currently see as likely.

More likely is a sequence of payloads to different addresses - in that case the ability to send all commands (possibly coils or registers) all within the same connect -> close session might be useful?

Reply:
I am not sure this is correct @march83 

```
    def write_coils(
        self,
        address: int,
        values: Union[List[bool], bool],
        slave: int = 0,
        **kwargs: Any,
    ) -> ModbusResponse:
        """Write coils (code 0x0F).

        :param address: Start address to write to
        :param values: List of booleans to write, or a single boolean to write
        :param slave: (optional) Modbus slave ID
        :param kwargs: (optional) Experimental parameters.
        :raises ModbusException:
        """
        return self.execute(
            pdu_bit_write.WriteMultipleCoilsRequest(address, values, slave, **kwargs)
        )
```

As you can see, the method is an step of indirection.

If you pay attention to the method signature there also:

```
class WriteMultipleCoilsRequest(ModbusRequest):
    """This function code is used to forcea sequence of coils.

    To either ON or OFF in a remote device. The Request PDU specifies the coil
    references to be forced. Coils are addressed starting at zero. Therefore
    coil numbered 1 is addressed as 0.

    The requested ON/OFF states are specified by contents of the request
    data field. A logical "1" in a bit position of the field requests the
    corresponding output to be ON. A logical "0" requests it to be OFF."
    """

    function_code = 15
    function_code_name = "write_coils"
    _rtu_byte_count_pos = 6

    def __init__(self, address=None, values=None, slave=None, **kwargs):
        """Initialize a new instance.

        :param address: The starting request address
        :param values: The values to write
        """
        ModbusRequest.__init__(self, slave=slave, **kwargs)
        self.address = address
        if not values:
```

I think if anything, we are doing the right thing for now.

Reply:
<img width="494" alt="image" src="https://github.com/EvergenEnergy/remote-comands-handler/assets/1817347/748ecc85-91e2-4b0a-aff6-31946b9d0b3d">

This are all the write methods published by the `client`. 

We probably may need to switch, from using the plain `client` abstraction, to build our own if this doesn't satisfy our needs, but I would recommend to do that as a follow up PR

Reply:
Yep, you are right. It's a list of coils with a start address.

This aligns with modbus function code 15, makes sense now.

Reply:
To your original question, if there is a list of non-consecutive coils to write to then they need to be handled independently with a new function call each time.